R/license_note.R

```diff
@@ -1,97 +1,120 @@
-        i = "Run {.code cargo install cargo-license} from your terminal."
-      ),
-      class = "rextendr_error"
-    )
-  }
+#'
+#' @examples
+#' \dontrun{
+#' write_license_note()
+#' }
+write_license_note <- function(
+    path = ".",
+    echo = TRUE,
+    quiet = FALSE,
+    overwrite = TRUE) {
+  check_string(path, class = "rextendr_error")
+  check_bool(echo, class = "rextendr_error")
+  check_bool(quiet, class = "rextendr_error")
+  check_bool(overwrite, class = "rextendr_error")
+
+  outfile <- rprojroot::find_package_root_file(
+    "LICENSE.note",
+    path = path
+  )
+
+  metadata <- read_cargo_metadata(
+    path = path,
+    dependencies = TRUE
+  )
 
-  manifest_file <- rprojroot::find_package_root_file("src", "rust", "Cargo.toml", path = path)
-  outfile <- rprojroot::find_package_root_file("LICENSE.note", path = path)
+  packages <- metadata[["packages"]]
 
-  if (!isTRUE(force) && file.exists(outfile)) {
-    cli::cli_abort(
-      c(
-        "LICENSE.note already exists.",
-        "If you want to regenerate LICENSE.note, set `force = TRUE` to {.fun write_license_note}."
-      ),
-      class = "rextendr_error"
-    )
+  # exclude current package from LICENSE.note
+  current_package <- metadata[["resolve"]][["root"]]
+
+  packages <- packages[packages[["id"]] != current_package, ]
+
+  replace_na <- function(data, replace = NA, ...) {
+    if (vctrs::vec_any_missing(data)) {
+      missing <- vctrs::vec_detect_missing(data)
+      data <- vctrs::vec_assign(data, missing, replace,
+        x_arg = "data",
+        value_arg = "replace"
+      )
+    }
+    data
   }
 
-  list_license <- processx::run(
-    "cargo",
-    c(
-      "license",
-      "--authors",
-      "--json",
-      "--avoid-build-deps",
-      "--avoid-dev-deps",
-      "--manifest-path", manifest_file
-    )
-  )$stdout %>%
-    jsonlite::parse_json()
-
-  package_names <- processx::run(
-    "cargo",
-    c(
-      "metadata",
-      "--no-deps",
-      "--format-version", "1",
-      "--manifest-path", manifest_file
+  packages[["respository"]] <- replace_na(
+    packages[["repository"]],
+    "unknown"
+  )
+
+  packages[["licenses"]] <- replace_na(
+    packages[["repository"]],
+    "not provided"
+  )
+
+  prep_authors <- function(authors, package) {
+    authors <- ifelse(
+      is.na(authors),
+      paste0(package, " authors"),
+      authors
     )
-  )$stdout %>%
-    jsonlite::parse_json() %>%
-    purrr::pluck("packages") %>%
-    purrr::map_chr("name")
-
-  .prep_authors <- function(authors, package) {
-    ifelse(!is.null(authors), authors, paste0(package, " authors")) %>%
-      stringi::stri_replace_all_regex(r"(\ <.+?>)", "") %>%
-      stringi::stri_replace_all_regex(r"(\|)", ", ")
+
+    authors <- stringi::stri_replace_all_regex(authors, r"(\ <.+?>)", "")
+
+    paste0(authors, collapse = ", ")
   }
 
+  packages[["authors"]] <- unlist(Map(
+    prep_authors,
+    packages[["authors"]],
+    packages[["name"]]
+  ))
```

Comment:
code cov wants us to check that `packages[["authors"]]` and `packages[["name"]]` is not null and the correct type. We can add some top level checks after `packages` is created with errors being thrown in the appropriate condition.